### ISSUE TITLE
can i using pretraining for some regression problems?

### ISSUE THREAD

**Issue by morestart** [agemagician/ProtTrans#107]
Can I use ProtTrans pretraining, add a fully connected layer, and change the loss to MSE for predicting some problems such as thermal stability?

**Comment by mheinzinger** [comment]
Sure, you can add a fully connected layer and a MSE loss to finetune the models.
You just need to be careful regarding hyperparameter choice s.a. learning-rate. You might also want to fine-tune only a subset of all layers. But all of this depends on your specific dataset and its size.
If you do not want to finetune but just want to train a regression predictor on top of existing, frozen embeddings, bio_trainer should make your life a lot easier: https://github.com/sacdallago/biotrainer

**Comment by morestart** [comment]
@mheinzinger I would like to know if the pool_strategy function is still required in regression tasks?
this is my code:
```Python
class BertWithRegression(pl.LightningModule):
    def __init__(
            self,
            vocab_size: int,
            nr_frozen_epochs: int,
            encoder_learning_rate: float,
            fine_tune_learning_rate: float
    ):
        super().__init__()
        self.model_name = "Rostlab/prot_bert_bfd"
        self.vocab_size = vocab_size
        self.encoder_learning_rate = encoder_learning_rate
        self.fine_tune_learning_rate = fine_tune_learning_rate
        self.nr_frozen_epochs = nr_frozen_epochs

        self.__build_model()

        if nr_frozen_epochs > 0:
            self.freeze_encoder()
        else:
            self._frozen = False

    def pool_strategy(self, features,
                      pool_cls=True, pool_max=True, pool_mean=True,
                      pool_mean_sqrt=True):
        token_embeddings = features['token_embeddings']
        cls_token = features['cls_token_embeddings']
        attention_mask = features['attention_mask']

        # Pooling strategy
        output_vectors = []
        if pool_cls:
            output_vectors.append(cls_token)
        if pool_max:
            input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
            token_embeddings[input_mask_expanded == 0] = -1e9  # Set padding tokens to large negative value
            max_over_time = torch.max(token_embeddings, 1)[0]
            output_vectors.append(max_over_time)
        if pool_mean or pool_mean_sqrt:
            input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
            sum_embeddings = torch.sum(token_embeddings * input_mask_expanded, 1)

            # If tokens are weighted (by WordWeights layer), feature 'token_weights_sum' will be present
            if 'token_weights_sum' in features:
                sum_mask = features['token_weights_sum'].unsqueeze(-1).expand(sum_embeddings.size())
            else:
                sum_mask = input_mask_expanded.sum(1)

            sum_mask = torch.clamp(sum_mask, min=1e-9)

            if pool_mean:
                output_vectors.append(sum_embeddings / sum_mask)
            if pool_mean_sqrt:
                output_vectors.append(sum_embeddings / torch.sqrt(sum_mask))

        output_vector = torch.cat(output_vectors, 1)
        return output_vector

    def __build_model(self):
        self.bert_model = BertModel.from_pretrained(self.model_name)
        self.encoder_features = 1024

        self.regression_head = nn.Sequential(
            nn.Linear(self.encoder_features * 4, 1024),
            nn.ReLU(),
            nn.Linear(1024, 1)
        )

    def unfreeze_encoder(self) -> None:
        """ un-freezes the encoder layer. """
        if self._frozen:
            logger.info(f"\n-- Encoder model fine-tuning")
            for param in self.ProtBertBFD.parameters():
                param.requires_grad = True
            self._frozen = False

    def freeze_encoder(self) -> None:
        """ freezes the encoder layer. """
        for param in self.ProtBertBFD.parameters():
            param.requires_grad = False
        self._frozen = True

    def forward(self, input_ids, token_type_ids, attention_mask):
        input_ids = torch.tensor(input_ids)
        attention_mask = torch.tensor(attention_mask)

        word_embeddings = self.bert_model(input_ids,
                                          attention_mask)[0]

        pooling = self.pool_strategy({"token_embeddings": word_embeddings,
                                      "cls_token_embeddings": word_embeddings[:, 0],
                                      "attention_mask": attention_mask,
                                      })
        out = self.regression_head(pooling)
        return out
```

**Comment by mheinzinger** [comment]
Technically: yes, you can drop the pooling and simply fine-tune based on the embedding of the CLS-token.
However, I would recommend to at least compare the different approaches because average pooling performed in our hands better so far.
What I would definitely do is to use ProtT5 and not ProtBERT. ProtT5 performed better throughout all our benchmarks (only use the encoder-part of ProtT5 and run it in half-precision; to simplify usage we created a separate model for this setup: https://huggingface.co/Rostlab/prot_t5_xl_half_uniref50-enc ).
For ProtT5, there won't be a CLS at the beginning but only one special token at the end of each sequence.
Depending on how much data you have, you can also consider to learn the pooling by inputting residue-embeddings to your regression model. See light-attention for an example on this: https://github.com/HannesStark/protein-localization/blob/master/models/light_attention.py

**Comment by morestart** [comment]
@mheinzinger Hello, mheinzinger. I found out that the max_length of the Rostlab/prot_bert_bfd model can only be set to 512. If I have a longer sequence, it will be truncated. Would this affect the final result, especially if the mutation site is after the 512th position?

**Comment by morestart** [comment]
I found in the official blog (https://ai.googleblog.com/2020/02/exploring-transfer-learning-with-t5.html) that when using the T5 model for training, it is recommended to use it to predict strings instead of the numbers themselves. I wonder if this also applies when using prot_t5_xl_half_uniref50-enc. I only have 3k data and the results are not very good when predicting directly using numbers.

**Comment by mheinzinger** [comment]
Yes, for T5 it is generally recommended to map your problem to string generation (yes, even if you want to predict classes/numbers) because T5 is a seq2seq model which is trained on generating output text.
That being said: I am not sure whether it makes sense to fine-tune such a large model if you only have 3k samples. In our hands this usually did not work well.
I would simply see how far you get with average pooling (simply sum over residue embeddings and divide by length) and then you put a normal regression on top of those embeddings.
If you want to go for fine-tuning, I would recommend this repo: https://github.com/r-three/t-few

**Comment by morestart** [comment]
I am certain that this method works, but I am not sure if it can be fine-tuned for such a large model. You can take a look at this notebook: https://www.kaggle.com/competitions/novozymes-enzyme-stability-prediction/discussion/361674

**Comment by mheinzinger** [comment]
I see.
So if you used only https://huggingface.co/Rostlab/prot_t5_xl_half_uniref50-enc , you should not need the conversion to letters as you stripped off the decoder/generator anyway. Maybe try fine-tuning only the last layer?

**Comment by morestart** [comment]
Yes, in the above wandb results, I only fine-tuned the regression head and trained for 30 epochs, but the results were relatively poor.

**Comment by mheinzinger** [comment]
I can not access your wandb because I have no access: "It looks like you’ve landed on a locked or empty page".
The snippet you copy&pasted above looks to me as if you commented out the unfreezing of the LM. If that is correct, you never fine-tune the LM but just train the regression on top of it.
From a quick glance at the github you shared above, it also looks to me as if you trained on sequences up to 1.5k length which is probably leading to OOM. 
In your script you also have a parameter after which epoch the LM is unfrozen. This is set to 5. This is why you experience OOM only after the fifth epoch, because only after this you start fine-tuning the whole encoder.
Try to finetune only the very last layer of the LM and try to lower the length of your input sequences. This way you should not run into OOM

**Comment by morestart** [comment]
Sorry, I set the permissions incorrectly. Here is the link that can be accessed: https://api.wandb.ai/links/cattree/thh4bmrw. I have a question, if the sequence is cropped to 512, is the data still valid if the mutation site is after 512? Regarding the OOM problem after epoch 5, I have solved it. I just need to freeze the model and not unfreeze it after the fifth epoch.

**Comment by morestart** [comment]
Thank you very much! I will recut the data according to your method and then post the results here.

**Comment by morestart** [comment]
Hello, mheinzinger. I tried using a smaller ESM model, specifically the esm2_t6_8M_UR50D. However, I found that when using a model with 600 million parameters, there is an order of magnitude difference between the validation loss and training loss, indicating overfitting. This is likely due to the limited amount of data I have, which consists of 4098 samples with an 8:2 ratio for training and validation. Here is the loss graph from Wandb: https://api.wandb.ai/links/cattree/5fu7xixl.
